Rubberduck.Core/AutoComplete/Service/AutoCompleteService.cs

```diff
@@ -0,0 +1,151 @@
+            try
+            {
+                if (!_initialized)
+                {
+                    var config = _configService.LoadConfiguration();
+                    ApplyAutoCompleteSettings(config);
+                }
+            }
+            finally
+            {
+                _initializing = false;
+            }
+        }
+
+        public void Enable()
+        {
+            if (!_initializing)
+            {
+                InitializeConfig();
+            }
+
+            if (!_enabled)
+            {
+                VBENativeServices.KeyDown += HandleKeyDown;
+                VBENativeServices.IntelliSenseChanged += HandleIntelliSenseChanged;
+                _enabled = true;
+            }
+        }
+
+        public void Disable()
+        {
+            if (_enabled && _initialized)
+            {
+                VBENativeServices.KeyDown -= HandleKeyDown;
+                VBENativeServices.IntelliSenseChanged -= HandleIntelliSenseChanged;
+                _enabled = false;
+            }
+        }
+
+        private void HandleIntelliSenseChanged(object sender, IntelliSenseEventArgs e)
+        {
+            _popupShown = e.Visible;
+        }
+
+        private void ConfigServiceSettingsChanged(object sender, ConfigurationChangedEventArgs e)
+        {
+            var config = _configService.LoadConfiguration();
+            ApplyAutoCompleteSettings(config);
+        }
+        
+        public void ApplyAutoCompleteSettings(Configuration config)
+        {
+            _settings = config.UserSettings.AutoCompleteSettings;
+            if (_settings.IsEnabled)
+            {
+                Enable();
+            }
+            else
+            {
+                Disable();
+            }
+            _initialized = true;
+        }
+
+        private bool WillHandle(AutoCompleteEventArgs e)
+        {
+            Debug.Assert(_settings != null);
+
+            if (!_enabled)
+            {
+                Debug.Assert(_enabled, "KeyDown controller is executing, but auto-completion service is disabled.");
+                return false; // release build should bail out here!
+            }
+
+            if (_popupShown || e.Character == default && e.IsDeleteKey)
+            {
+                return false;
+            }
+
+            var module = e.Module;
+            using (var pane = module.CodePane)
+            {
+                if (pane.Selection.LineCount > 1)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private void HandleKeyDown(object sender, AutoCompleteEventArgs e)
+        {
+            if (!WillHandle(e))
+            {
+                return;
+            }
+
+            foreach (var handler in _handlers)
+            {
+                var result = handler.Handle(e, _settings);
+                if (result != null && e.Handled)
```

Comment:
I already criticised this on the now-closed PR: this looks like it wants to be a `TryHandle`. The handler shouldn't set the handled state of the ACEvent